### PR TITLE
FRR: block incoming routes

### PR DESCRIPTION
**Pull request by fedepaol**
BGP routes must only be published. We should not accept incoming routes, as FRR is host networked and that would result in unwanted routes at host level of the cluster.

We use prefix-list to block the routes.

### CODE REVIEW

File: internal/bgp/frr/config.go [hunk 1]
```diff
@@ -66,7 +69,9 @@ router bgp {{.MyASN}}
   {{- end }}
 {{- if ne .BFDProfile ""}} 
   neighbor {{.Addr}} bfd profile {{.BFDProfile}}
-{{- end -}}
+{{- end }}
+  neighbor {{.Addr}} prefix-list deny-all in
```
Comment (msherif1234) [inline internal/bgp/frr/config.go:73]: nit v4deny-all

File: internal/bgp/frr/config.go [hunk 2]
```diff
@@ -66,7 +69,9 @@ router bgp {{.MyASN}}
...
+  neighbor {{.Addr}} prefix-list v6deny-all in
```
Comment (msherif1234) [inline internal/bgp/frr/config.go:74]: this should be done inside address-family block where we apply route-map and based on the version u pick one or the other

**Comment by russellb** [comment]
Lgtm

**Comment by msherif1234** [comment]
so 0.0.0.0/0 and 0::0/0 came to FRR speaker because FRR test containers are host network ? what if we have real deployment and the other side have all kind of routes static or other routing protocols is there a way to block all none BGP routes ?

**Comment by fedepaol** [comment]
> so 0.0.0.0/0 and 0::0/0 came to FRR speaker because FRR test containers are host network ? what if we have real deployment and the other side have all kind of routes static or other routing protocols is there a way to block all none BGP routes ?

0.0.0.0/0 and the other one are a way to block ALL the addresses. By doing this we are blocking all the routes coming from the other side.
If you look at the test, I tried to inject a 32 address, and if you remove this code that route will be added to the host (because frr runs hostnetworked).

File: e2etest/bgp_test.go [hunk 3]
```diff
@@ -946,6 +946,66 @@ var _ = ginkgo.Describe("BGP", func() {
...
+						return fmt.Errorf("Found %s in %s routes", toInject, pod.Name)
+					}
+				}
+			}
+			return nil
+		}, 30*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
+
+	},
+		table.Entry("IPV4", "192.168.10.0/24", "ipv4", "172.16.1.10/32"),
+		table.Entry("IPV6", "fc00:f853:0ccd:e799::/116", "ipv6", "fc00:f853:ccd:e800::1/128"),
```
Comment (msherif1234) [inline e2etest/bgp_test.go:1006]: those /32 and 128 are the NIC IP address can we use const and name them accordingly to make obvious ?

**Comment by msherif1234** [comment]
> > so 0.0.0.0/0 and 0::0/0 came to FRR speaker because FRR test containers are host network ? what if we have real deployment and the other side have all kind of routes static or other routing protocols is there a way to block all none BGP routes ?
> 
> 0.0.0.0/0 and the other one are a way to block ALL the addresses. By doing this we are blocking all the routes coming from the other side. If you look at the test, I tried to inject a 32 address, and if you remove this code that route will be added to the host (because frr runs hostnetworked).

so "0" means "*" wildcard in this prefix-list

File: e2etest/bgp_test.go (hunk 3 above)
Comment (fedepaol) [reply e2etest/bgp_test.go:1006]: They are not the NIC ips. They are random ips that I am injecting routes for.
The corresponding name is the function parameter, I can change that to make it more clear.

File: internal/bgp/frr/config.go (hunk 2 above)
Comment (fedepaol) [reply internal/bgp/frr/config.go:74]: This will make the template uglier, as we'd need to do it both when there are no advertisements and when we have those, plus iff-ing on the ip family.
I think it's fair to say we always block everything.

File: e2etest/bgp_test.go (hunk 3 above)
Comment (msherif1234) [reply e2etest/bgp_test.go:1006]: cool I wrote that comment before u replied to other one so its clear to me know

File: internal/bgp/frr/config.go (hunk 2 above)
Comment (msherif1234) [reply internal/bgp/frr/config.go:74]: ugly ofcourse but it address-family config, not sure what will happen when u pass v4 address to v6 prefix list and other way around to be safe having them under their respective address-family will be less error prune IMO

Comment (fedepaol) [reply internal/bgp/frr/config.go:74]: (I mean, if it works)

Comment (fedepaol) [reply internal/bgp/frr/config.go:74]: I think you are right. The resulting configuration renders like this :

```
 !
 address-family ipv4 unicast
  neighbor fc00:f853:ccd:e793::5 activate
  neighbor fc00:f853:ccd:e793::5 prefix-list v6deny-all in
  neighbor fc00:f853:ccd:e793::6 activate
  neighbor fc00:f853:ccd:e793::6 prefix-list v6deny-all in
 exit-address-family
 !
 address-family ipv6 unicast
  neighbor fc00:f853:ccd:e793::5 activate
  neighbor fc00:f853:ccd:e793::6 activate
 exit-address-family
```

File: internal/bgp/frr/config.go [hunk 4]
```diff
@@ -44,6 +44,9 @@ route-map {{$n.Addr}}-map permit 10
 {{- end }}
 {{- end }}
 
+ip prefix-list deny-all deny 0.0.0.0/0
+ip prefix-list v6deny-all deny ::/0
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply internal/bgp/frr/config.go:48]: I think I tried that and resulted in an error. Let me try again.

[the remark this answers is not in the document]
Comment (fedepaol) [reply internal/bgp/frr/config.go:48]: Yep that yelds:
```
b'line 9: % Unknown command: ip6 prefix-list v6deny-all deny ::/0\n\n'
```

[the remark this answers is not in the document]
Comment (cgoncalves) [reply internal/bgp/frr/config.go:48]: Typo? Should be `ipv6`, not `ip6`

[the remark this answers is not in the document]
Comment (fedepaol) [reply internal/bgp/frr/config.go:48]: Ah, that works! Thanks for spotting it :-)

**Comment by fedepaol** [comment]
@msherif1234 @cgoncalves I think I addressed all comments here.

**Comment by markkrj** [comment]
> 0.0.0.0/0 and the other one are a way to block ALL the addresses. By doing this we are blocking all the routes coming from the other side. If you look at the test, I tried to inject a 32 address, and if you remove this code that route will be added to the host (because frr runs hostnetworked).

If you permit a  comment about this from an outsider, this will work because FRR (and all routers that I know of) assume an implicit deny on prefix-lists (and ACLs in general), but your rule is not getting matched at all... 0.0.0.0/0 matches only this exact prefix (a default route). To match against any prefix, you must use "0.0.0.0/0 le 32".

**Comment by cgoncalves** [comment]
Good point. To make it even more explicit, consider "deny any"
```
centos8.localdomain(config)# ip prefix-list deny-all deny any
centos8.localdomain(config)# ipv6 prefix-list v6deny-all deny any
centos8.localdomain(config)# exit
centos8.localdomain# show running-config 
Building configuration...

Current configuration:
!
frr version 7.5
frr defaults traditional
hostname centos8.localdomain
no ip forwarding
no ipv6 forwarding
no service integrated-vtysh-config
!
ip prefix-list deny-all seq 5 deny any
!
ipv6 prefix-list v6deny-any seq 5 deny any
!
line vty
!
end
```

**Comment by fedepaol** [comment]
Thanks @markkrj and @cgoncalves . I'll use the `any` version, seems more readable to me.

File: internal/bgp/frr/config.go [hunk 5]
```diff
@@ -66,21 +69,24 @@ router bgp {{.MyASN}}
...
   neighbor {{.Addr}} bfd profile {{.BFDProfile}}
-{{- end -}}
+{{- end }}
 {{- end }}
 {{range $n := .Neighbors -}}
 {{/* no bgp default ipv4-unicast prevents peering if no address families are defined. We declare an ipv4 one for the peer to make the pairing happen */}}
 {{- if eq (len .Advertisements) 0}}
   address-family ipv4 unicast
     neighbor {{$n.Addr}} activate
+    neighbor {{.Addr}} prefix-list ipv4deny-all in
```
Comment (ton31337) [inline internal/bgp/frr/config.go:79]: I'd recommend using route-map for in/out route filtering.

```
route-map {{.Addr}}-in permit {{.index}}
  match ip address prefix-list ipv4deny-all
route-map {{.Addr}}-in permit {{.index}}
  match ipv6 address prefix-list ipv6deny-all
...
neighbor {{.Addr}} route-map {{.Addr}}-in in
neighbor {{.Addr}} route-map {{.Addr}}-out out
```

Just will be more flexible in the future.

Comment (fedepaol) [reply internal/bgp/frr/config.go:79]: will do, thanks!

**Comment by msherif1234** [comment]
I have question/doubt with this rule in place and the other end has BGP redistribute config for any IGP routes will this rule now incorrectly drop those routes ?

**Comment by fedepaol** [comment]
> I have question/doubt with this rule in place and the other end has BGP redistribute config for any IGP routes will this rule now incorrectly drop those routes ?

not sure I understood. On metallb side we must drop all the incoming routes. My understanding is that htey affect only the local instance.

**Comment by cgoncalves** [comment]
FRR-based speakers will deny installing in its RIB any IP prefixes advertised by its BGP neighbors. In other words, it will receive the prefixes but drop them (think of it as an iptables DROP rule).

File: internal/bgp/frr/config.go (hunk 5 above)
Comment (msherif1234) [reply internal/bgp/frr/config.go]: @ton31337 can FRR handle multiple route-map applied in the same direction or they have to be merged into single route-map ?

Comment (fedepaol) [reply internal/bgp/frr/config.go]: I am not sure this version looks more readable, I may revert and bring in route-map(s) when we need to deal with more complex scenarios.
Thoughts?

**Comment by cgoncalves** [comment]
Personally, while I agree that @ton31337's suggestion provides more flexibility, it is not something we require today and can be easily changed later. I am also not sure how it would scale with a large number of neighbors as it would result in having a route-map, each with its own prefix-list, for each neighbor.

My preference goes to https://github.com/metallb/metallb/pull/1100#issuecomment-986700928 but I am definitely biased here :sweat_smile:

File: internal/bgp/frr/config.go (hunk 5 above)
Comment (msherif1234) [reply internal/bgp/frr/config.go]: fortunately the existing route-map in the "out" direction so we don't have to merge yet :) so I prefer using route-map I understand the template becoming hard to read but I know from day1 this where it will go I will revive some initial attempt to make it a hierachy of templates not only one as it won't scale with bgp configs

**Comment by msherif1234** [comment]
> Personally, while I agree that @ton31337's suggestion provides more flexibility, it is not something we require today and can be easily changed later. I am also not sure how it would scale with a large number of neighbors as it would result in having a route-map, each with its own prefix-list, for each neighbor.
> 
> My preference goes to [#1100 (comment)](https://github.com/metallb/metallb/pull/1100#issuecomment-986700928) but I am definitely biased here sweat_smile

for scale we probably will need to adopt peer-group configuration

**Comment by ton31337** [comment]
>for scale we probably will need to adopt peer-group configuration

Also, consider using dynamic peers as well (if talking about the scale) :)

File: internal/bgp/frr/config.go (hunk 5 above)
Comment (fedepaol) [reply internal/bgp/frr/config.go]: Let's try to wrap up.
We can have only one route-map per family per direction per neighbor? 
Like here it's allowed having two:

```
route-map {{.Addr}}-in permit {{.index}}
  match ip address prefix-list ipv4deny-all
route-map {{.Addr}}-in permit {{.index}}
  match ipv6 address prefix-list ipv6deny-all
``` 
because they are for different ip families?

And scaling aside, what's the reason for going route-maps now and not later (as this will always be block every route in)?

Comment (ton31337) [reply internal/bgp/frr/config.go]: It was just a suggestion :) 

>as this will always be block every route in

What do you mean? Which example?

Comment (fedepaol) [reply internal/bgp/frr/config.go]: Was trying to "talk out loudly". What I meant is, what we are trying to achieve here is to always block incoming routes. Due to the nature of metallb, that is really unlikely to change so it's not that bad pinning it to a neighbor (I think).

**Comment by fedepaol** [comment]
Last version, as discussed with @ton31337 this is way more compact.
I also renamed the -in / -out rules to make it more clear.

File: internal/bgp/frr/testdata/TestBFDProfileAllDefault.golden [hunk 6]
```diff
@@ -3,6 +3,9 @@ log file /etc/frr/frr.log informational
 log timestamp precision 3
 hostname dummyhostname
 
+ip prefix-list ipv4deny-all deny any
```
Comment (cgoncalves) [inline internal/bgp/frr/testdata/TestBFDProfileAllDefault.golden:6]: Leftovers

Comment (fedepaol) [reply internal/bgp/frr/testdata/TestBFDProfileAllDefault.golden]: Thanks, fixed now

**Comment by cgoncalves** [comment]
> Last version, as discussed with @ton31337 this is way more compact. I also renamed the -in / -out rules to make it more clear.

Could you please help us understand why having a route-map in per neighbor was still preferred over prefix-list? Potential future flexibility over readability?

**Comment by ton31337** [comment]
>Could you please help us understand why having a route-map in per neighbor was still preferred over prefix-list? Potential future flexibility over readability?

@cgoncalves there is no such as _preferred_, it's just more like _universal_. From what I see usually, people prefer using route-maps everywhere and all the logic inside. Because now you have prefix-list and for some reason, you start using route-map, you will end up with two clashing filters (route-map + prefix-list) while you can have only a single one from the beginning. 

I'm not saying it's wrong as it is (with prefix-list), I just suggesting :) Both work fine. 

At least if you decide to drop everything except some routes for example tagged with arbitrary communities or so, route maps must be used. They are just flexible as you said for the future _features_ :)

**Comment by msherif1234** [comment]
> > for scale we probably will need to adopt peer-group configuration
> 
> Also, consider using dynamic peers as well (if talking about the scale) :)

using peer x.x.x.x/24 ? I didn't know FRR support that ? very cool !!

**Comment by cgoncalves** [comment]
LGTM
